### PR TITLE
8310530: PipedOutputStream.flush() accesses sink racily

### PR DESCRIPTION
**Pull request by stsypanov**
Just a tiny clean-up to remove racy read within synchronized method

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8310530](https://bugs.openjdk.org/browse/JDK-8310530): PipedOutputStream.flush() accesses sink racily (**Enhancement** - P4)

### Reviewers
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**)
 * [Brian Burkhalter](https://openjdk.org/census#bpb) (@bplb - **Reviewer**)
 * [Chen Liang](https://openjdk.org/census#liach) (@liach - Author)
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14589/head:pull/14589` \
`$ git checkout pull/14589`

Update a local copy of the PR: \
`$ git checkout pull/14589` \
`$ git pull https://git.openjdk.org/jdk.git pull/14589/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14589`

View PR using the GUI difftool: \
`$ git pr show -t 14589`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14589.diff">https://git.openjdk.org/jdk/pull/14589.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14589#issuecomment-1600908764)

### CODE REVIEW

File: src/java.base/share/classes/java/io/PipedOutputStream.java [hunk 1]
```diff
@@ -163,6 +163,7 @@ public void write(byte[] b, int off, int len) throws IOException {
      */
     @Override
     public synchronized void flush() throws IOException {
+        PipedInputStream sink = this.sink;
```
Comment (liach) [inline src/java.base/share/classes/java/io/PipedOutputStream.java:166]:
```suggestion
        var sink = this.sink;
```
As seen in other methods.

Comment (liach) [reply src/java.base/share/classes/java/io/PipedOutputStream.java:166]: On second thought, this is probably not necessary; write to the sink field is in another synchronized method, and this method is synchronized already. Is the goal here to remove the synchronized on flush?

Comment (stsypanov) [reply src/java.base/share/classes/java/io/PipedOutputStream.java]: I think it's better to keep `synchronized` over the method this far

**Comment by bplb** [comment]
@stsypanov The description section of the issue JDK-8310530 should be filled out. Thanks.

**Review by dfuch — APPROVED** [review]

**Review by bplb — APPROVED** [review]
Change looks fine.

**Comment by stsypanov** [comment]
@bplb done

**Comment by bplb** [comment]
> @bplb done

Thanks!

**Comment by RogerRiggs** [comment]
I think the bug description should include a bit more detail along the lines that since the method reading and writing are synchronized, its not really a racy read. It does avoid reading the volatile value twice (a relatively expensive operation).

**Review by liach — APPROVED** [review]

**Comment by stsypanov** [comment]
@RogerRiggs updated

**Review by RogerRiggs — APPROVED** [review]

**Comment by bplb** [comment]
I will sponsor this on Monday if it is still needed.

**Pull request closed without merging**